### DISCUSSION TITLE
Event Map Sortie logic [Partially Flandolly kcauto related]

### DISCUSSION THREAD

**Discussion by Synexcu** [XVs32/kcauto_custom#69] [Q&A] [answered]
Hello,

First of all, I would like to apologize if you have a hard time understanding my question, English is my 2nd language and It is sometimes hard for me to construct easy to understand words for my question compared to when I am using my 1st language.

So as the title have mentioned, I have a question regarding Event Map Sortie logic, partially Flandolly kcauto related. I may use some or a lot of scenarios to express my question so bear with me.

Yesterday, I tried making a file config for E5 for farming purposes in Flandolly version's of kcauto:
![kcautoE5](https://github.com/XVs32/kcauto_custom/assets/82300898/ab3bafdc-a02c-437a-8bf6-053f8256a91b)
All the usual, nothing's wrong.

I also replaced the event assets from the original to your version of the image from the `_event_next_page.png`  until the `_event_world_.png` 

But when I start running the program, it can not choose the correct map (E5) to sortie. It goes to the E7 in Page 4 of the event like this (see down below) whereas it is supposed to navigate to the E5 Page 3 of the event map:

![E5choose (1)](https://github.com/XVs32/kcauto_custom/assets/82300898/d665d84b-58bf-4de7-8800-8f2992294770)

Seeing this, I tried to 'map' it and look what caused this. When I set the program to sortie to E4, the program goes to Page 3; E3 to Page 2; E1 and E2 to Page 1.

![image](https://github.com/XVs32/kcauto_custom/assets/82300898/ae363713-0604-47b9-b5a8-51dd4a765733)

I looked a bit into the code, and I noticed this particular part in the `combat_core.py`:

```
    def _select_event_map(self, sortie_map):
        cur_page = 1
        if sortie_map.world == 'E':
            while cur_page < self.map_data.page:
                kca_u.kca.r['top'].hover()
                kca_u.kca.sleep(0.5)
                kca_u.kca.click_existing(
                    'kc', f'combat|_event_next_page_{cur_page}.png')
                cur_page += 1
        kca_u.kca.r['top'].hover()
        kca_u.kca.click_existing(
            'kc', f'combat|_event_world_{sortie_map.value}.png')
```

Why is the program not navigating E2 to Page 2? When all E3, E4, E5 has their own page for 1 map (E3 1 whole page, E4 1 whole page, and so on.) What caused this? Is it the `map_data.page` ? Or is it perhaps because of the previous developer coded the event mapping similar to the past event of Summer 2021 where there is exactly E1 and E2 in the Page 1 and E3 in the Page 2 which then would prove the program's action above but I don't have any concrete proof on the code to support this but theoretically based on my limited understanding of the whole program.

> **(Summer 2021 E1 and E2 in Page 1 image)**
![image](https://github.com/XVs32/kcauto_custom/assets/82300898/eefe1ca7-50aa-47e8-bcc9-c889ce29b15a)

> **(Summer 2021 E3 in Page 2 image)**
![image](https://github.com/XVs32/kcauto_custom/assets/82300898/f0ded8c2-7b3b-4551-bc02-c55516e40948)

If yes, how can I fix it so that it can read the event map and map the maps accordingly?

Thank you, I appreciate your time and attention.

**Comment by XVs32** [comment, marked as answer]
Hi [Synexcu](https://github.com/Synexcu),

There are a few things relate to page switching in kcauto:
* The assets image
* The page switch logic in `combat_core.py`
* The page setting in `E-X.json`

First, your kcauto was able to switch to the last page (page for E7), so I believe your assets image is good.
Second, Flandolly doesn't seems to changed any of the `combat_core.py`'s code, which means the page switch logic is probably fine(as it is for the pass 4 years).
Therefore, I would suggest you take a look into `data/combat/E-X.json`, especially `E-5.json`

In every `E-X.json`, there is a page element which defines which page to go:
```
{
    "world": "E",
    "subworld": 5,
    "panel": 1,
    "page": 3,            //This one 
...
}
```

Sincerely,
XVs32

**Comment by Synexcu** [comment]
Alright, I have updated the Event map combat data from E1 until E7. I have ran several tests for the E5 and yes, it worked the 1st and the 2nd loop as in it have correctly chose the map from the page. (More detail problem about this further down below)

Now I encountered another problem inside the map in combat where in E5 P node (submarine) I have selected to use the Line Abreast in the config, but the program chose Formation 4 (Line Ahead), although it only occurred once throughout the whole testing run, I think it is worth mentioning.

Next, when I tried running the program again, creating a new running instance different from the previous one, it failed selecting the E5 map and chose the E6 instead,

After knowing this, I decided to try to sortie to other event maps by setting it to the program, such as E6 and yes, it failed:

![E6choose (1)](https://github.com/XVs32/kcauto_custom/assets/82300898/ada0d11a-f270-4ff5-9a69-62ada4e7e9bf)

In another instance also, instead of going into Page 3, it goes to the Page 2 and then goes back to Page 1 and chose E2.

I tried running E3 in the program now, and it goes to the Page 2 instead of in Page 1 just as seen below:

![E3choose-min](https://github.com/XVs32/kcauto_custom/assets/82300898/0e609c55-87eb-4a32-a435-57fbe4cc74bd)

So, I looked into your `E-3.json` code in `data/combat/E-3.json` to debug and found that you wrote:

```
{
    "world": "E",
    "subworld": 3,
    "panel": 1,
    "page": 2, ..
```

I tried fixing the code by changing the `"panel"` and the `"page"` to this:
```
{
  "world": "E",
  "subworld": 3,
  "panel": 3,
  "page": 1, ..
```
and it worked fine and it successfully located the E3 map.

As for Page 3, for my case, it can't detect the `_event_next_page_2.png` asset, so it is stuck in Page 2 and clicked the E4 instead. Before, I managed to navigate to E5 because I have created my own `_event_next_page_2.png` asset because it didn't work. Once I changed it to my version, it worked perfectly.

> **`_event_next_page_2.png`**
![_event_next_page_2](https://github.com/XVs32/kcauto_custom/assets/82300898/7b812c07-9e96-4592-a907-9d34132b423f)

For Page 4, again it did not detect the  `_event_next_page_3.png` asset to sortie to map E7. What I did was, again, adding my own asset so that it would detect the button

> **`_event_next_page_3.png`**
![_event_next_page_3](https://github.com/XVs32/kcauto_custom/assets/82300898/d8c3065b-0812-4a94-8766-73409d730b35)

I tweaked some Event map combat data and added `"panel"` and `"page"` in E2 and E3,
And as for E4, I changed the from what previously was:
```
{
    "world": "E",
    "subworld": 4,
    "panel": 1,
    "page": 3, ..
```
into:
```
{
  "world": "E",
  "subworld": 4,
  "panel": 1,
  "page": 2, ..
```
because I am afraid it may overlap with E5's combat map data

> E5 combat map data

```
{
    "world": "E",
    "subworld": 5,
    "panel": 1,
    "page": 3,
```

> E4 combat map data

```
{
    "world": "E",
    "subworld": 4,
    "panel": 1,
    "page": 3,
```

E5 combat map data stays the same with your version.

Aside from this, I also would like to mention that every time the program attempts to navigate to the map selected, it sometimes clicked the area of the map which shares some pixel area for the button, hence sometimes it accidentally navigated to the wrong page and selecting the wrong map, is there a fix for this one perhaps?

Thanks for your time, I'm looking forward to your answer!

**Reply by Synexcu** [reply]
I forgot to add that it sometimes stuck choosing node like in here R>V :
![image](https://github.com/XVs32/kcauto_custom/assets/82300898/99924d22-a1c5-440f-863d-ef799e3106f4)

**Comment by XVs32** [comment]
As I said in my release note, I am stuck in E1 so only E1 is supported at the moment.

About your questions:
1. Mis-clicking to other page:  
kcauto clicks on the 'event_map_E-X' you provided, recapture those assets might solve the problem
2. Stuck in node selecting:
This is a network delay issue, I have noticed it a while ago, though it takes a combat logic re-do to fix.  
I need to take my time and do this carefully(since any bug in combat logic would cause ship lose).


After all, I did update E1-E7 'E-X.json' but those are not properly tested and it is there for those veterans(who are familiar with kcauto_custom).  
I am not willing to solve problems in E2-E7 until I clear E1(Since I don't have a proper env to debug).
You might just have to wait for my next release.
